### PR TITLE
Unblock PIP package hashes when logging models

### PR DESCRIPTION
**Pull request by maitre-matt**
Signed-off-by: Matthieu Maitre <mmaitre@microsoft.com>

## What changes are proposed in this pull request?

This addresses #4886. `_is_mlflow_requirement()` was not able to detect that `mlflow` had already been listed in PIP requirements when requirements include package hashes. Per https://github.com/pypa/packaging/issues/488, `Requirement` does not handle package hashes because this is a PIP-specific extension. The caller is responsible for removing, which is what this change does.

The change also adds an example showing how to manage package hashes using `pip-tools`.

## How is this patch tested?

Expanded existing unit test

## Does this PR change the documentation?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Make sure the changed pages / sections render correctly by following the steps below.

1. Check the status of the `ci/circleci: build_doc` check. If it's successful, proceed to the
   next step, otherwise fix it.
2. Click `Details` on the right to open the job page of CircleCI.
3. Click the `Artifacts` tab.
4. Click `docs/build/html/index.html`.
5. Find the changed pages / sections and make sure they render correctly.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

This adds support for the PIP hash-checking mode in ML projects.

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [x] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [x] `area/docs`: MLflow documentation pages
- [x] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [x] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [x] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: examples/supply_chain_security/train.py [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+# Load data
+X, y = sklearn.datasets.load_diabetes(return_X_y=True)
+X_train, X_test, y_train, y_test = sklearn.model_selection.train_test_split(
+    X, y, test_size=0.2, random_state=0
+)
+
+# Train model
+with mlflow.start_run() as run:
+    print(f"MLFlow run ID: {run.info.run_id}")
```
Comment (dbczumar) [inline examples/supply_chain_security/train.py:15]: Nit: 

```suggestion
    print(f"MLflow run ID: {run.info.run_id}")
```

**Review by dbczumar — APPROVED** [review]
LGTM! This is fantastic! Thank you for your contribution, @maitre-matt !

File: examples/supply_chain_security/README.rst [hunk 2]
```diff
@@ -0,0 +1,87 @@
+Python Package Anti-Tampering with MLflow
```
Comment (harupy) [inline examples/supply_chain_security/README.rst:1]: Can we include this example in `tests/`

File: mlflow/utils/environment.py [hunk 3]
```diff
@@ -230,6 +231,18 @@ def _is_mlflow_requirement(requirement_string):
     except InvalidRequirement:
         # A local file path or URL falls into this branch.
 
+        # `Requirement` throws an `InvalidRequirement` exception if `requirement_string` contains per-requirement options (ex: package hashes)
+        # GitHub issue: https://github.com/pypa/packaging/issues/488
+        # Per-requirement-option spec: https://pip.pypa.io/en/stable/reference/requirements-file-format/#per-requirement-options
+        if "--" in requirement_string:
+            match = re.search("^(.*?)((?:\\s+--\\w+=\\S+)*)$", requirement_string)
```
Comment (harupy) [inline mlflow/utils/environment.py:238]:
```suggestion
            match = re.search(r"^(.*?)((?:\s+--\w+=\S+)*)$", requirement_string)
```

Can we use a raw string?

Comment (harupy) [reply mlflow/utils/environment.py:238]: Found a function used in pip for processing a requirement string:
https://github.com/pypa/pip/blob/b392833a0f1cff1bbee1ac6dbe0270cccdd0c11f/src/pip/_internal/req/req_file.py#L400

```python
from typing import Tuple


def break_args_options(line: str) -> Tuple[str, str]:
    """Break up the line into an args and options string.  We only want to shlex
    (and then optparse) the options, not the args.  args can contain markers
    which are corrupted by shlex.
    """
    tokens = line.split(" ")
    args = []
    options = tokens[:]
    for token in tokens:
        if token.startswith("-") or token.startswith("--"):
            break
        else:
            args.append(token)
            options.pop(0)
    return " ".join(args), " ".join(options)


print(break_args_options("mlflow"))
print(break_args_options("mlflow==1.2.3"))
print(break_args_options("-r reqs.txt"))
print(break_args_options("mlflow==1.2.3 --hash=foo"))
```

output:

```
('mlflow', '')
('mlflow==1.2.3', '')
('', '-r reqs.txt')
('mlflow==1.2.3', '--hash=foo')
```

Can we do something similar?

File: tests/utils/test_environment.py [hunk 4]
```diff
@@ -244,6 +244,23 @@ def test_process_pip_requirements(tmpdir):
...
 
+    # Ensure a requirement for mlflow is preserved when package hashes are specified
+    conda_env, reqs, cons = _process_pip_requirements(
+        ["a"],
+        pip_requirements=[
+            """mlflow==1.20.2 \
+    --hash=sha256:963c22532e82a93450674ab97d62f9e528ed0906b580fadb7c003e696197557c \
+    --hash=sha256:b15ff0c7e5e64f864a0b40c99b9a582227315eca2065d9f831db9aeb8f24637b"""
+        ],
+    )
```
Comment (harupy) [inline tests/utils/test_environment.py:255]:
```suggestion
    hash1 = "963c22532e82a93450674ab97d62f9e528ed0906b580fadb7c003e696197557c"
    hash2 = "b15ff0c7e5e64f864a0b40c99b9a582227315eca2065d9f831db9aeb8f24637b"
    conda_env, reqs, cons = _process_pip_requirements(
        ["a"], pip_requirements=[f"mlflow==1.20.2 --hash=sha256:{hash1} --hash=sha256:{hash2}"],
    )
```

Can we store hashes in variables to improve the readability?

File: examples/supply_chain_security/train.py (hunk 1 above)
Comment (mmaitre314) [reply examples/supply_chain_security/train.py:15]: Fixed the typo

File: tests/utils/test_environment.py (hunk 4 above)
Comment (mmaitre314) [reply tests/utils/test_environment.py:255]: Test refactored

File: mlflow/utils/environment.py (hunk 3 above)
Comment (mmaitre314) [reply mlflow/utils/environment.py:238]: Updated code to `split`-based solution

File: examples/supply_chain_security/README.rst (hunk 2 above)
Comment (maitre-matt) [reply examples/supply_chain_security/README.rst:1]: Included the example in the test

File: tests/utils/test_environment.py [hunk 5]
```diff
@@ -244,6 +256,20 @@ def test_process_pip_requirements(tmpdir):
...
+    conda_env, reqs, cons = _process_pip_requirements(
+        ["a"], pip_requirements=[f"mlflow==1.20.2 --hash={hash1} --hash={hash2}"],
+    )
+    assert _get_pip_deps(conda_env) == [
+        "mlflow==1.20.2 --hash=sha256:963c22532e82a93450674ab97d62f9e528ed0906b580fadb7c003e696197557c --hash=sha256:b15ff0c7e5e64f864a0b40c99b9a582227315eca2065d9f831db9aeb8f24637b"
+    ]
+    assert reqs == [
+        "mlflow==1.20.2 --hash=sha256:963c22532e82a93450674ab97d62f9e528ed0906b580fadb7c003e696197557c --hash=sha256:b15ff0c7e5e64f864a0b40c99b9a582227315eca2065d9f831db9aeb8f24637b"
+    ]
+    assert cons == []
```
Comment (harupy) [inline tests/utils/test_environment.py:271]: Can we use `hash1` and `hash2` here?

Comment (maitre-matt) [reply tests/utils/test_environment.py]: Replaced hashes with `hash1`/`hash2`

**Review by harupy — APPROVED** [review]
@maitre-matt Thanks, this is a nice enhancement!

**Pull request merged**